lib/os/onoff.c

```diff
@@ -0,0 +1,522 @@
+		break;
+	case SERVICE_STATE_TO_ON:
+		/* Already starting, just queue it */
+		add_client = true;
+		rv = 1;
+		break;
+	case SERVICE_STATE_ON:
+		/* Just increment the reference count */
+		notify = true;
+		break;
+	default:
+		rv = -EINVAL;
+		break;
+	}
+
+out:
+	if (add_client) {
+		sys_slist_append(&srv->clients, &cli->node);
+	} else if (notify) {
+		srv->refs += 1;
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	if (start) {
+		__ASSERT_NO_MSG(srv->start != NULL);
+		srv->start(srv, onoff_start_notify);
+	} else if (notify) {
+		notify_one(srv, cli, 0);
+	}
+
+	return rv;
+}
+
+static void onoff_stop_notify(struct onoff_service *srv,
+			      int res)
+{
+	bool notify_clients = false;
+	int client_res = res;
+	bool start = false;
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+	struct onoff_client *releaser = srv->releaser;
+
+	/* If the stop operation failed log an error and leave the
+	 * rest of the state in place.
+	 *
+	 * If it succeeded remove the last reference and transition to
+	 * off.
+	 *
+	 * In either case remove the last reference, and notify all
+	 * waiting clients of operation completion.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+		srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+		notify_clients = true;
+	} else if (sys_slist_is_empty(&clients)) {
+		set_service_state(srv, SERVICE_STATE_OFF);
+	} else if ((k_is_in_isr() || k_is_pre_kernel())
+		   && ((srv->flags & ONOFF_SERVICE_START_SLEEPS) != 0U)) {
+		set_service_state(srv, SERVICE_STATE_OFF);
+		notify_clients = true;
+		client_res = -EWOULDBLOCK;
+	} else {
+		set_service_state(srv, SERVICE_STATE_TO_ON);
+		start = true;
+	}
+
+	__ASSERT_NO_MSG(releaser);
+	srv->refs -= 1U;
+	srv->releaser = NULL;
+	__ASSERT_NO_MSG(srv->refs == 0);
+
+	/* Remove the clients if there was an error or a delayed start
+	 * couldn't be initiated, because we're resolving their
+	 * operation with an error.
+	 */
+	if (notify_clients) {
+		sys_slist_init(&srv->clients);
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	/* Notify the releaser.  If there was an error, notify any
+	 * pending requests; otherwise if there are pending requests
+	 * start the transition to ON.
+	 */
+	notify_one(srv, releaser, res);
+	if (notify_clients) {
+		notify_all(srv, &clients, client_res);
+	} else if (start) {
+		srv->start(srv, onoff_start_notify);
+	}
+}
+
+int onoff_release(struct onoff_service *srv,
+		  struct onoff_client *cli)
+{
+	bool stop = false;      /* invoke stop transition */
+	bool notify = false;    /* do client notification */
+	int rv = validate_args(srv, cli);
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) != 0) {
+		rv = -EIO;
+		goto out;
+	}
+
+	u32_t state = srv->flags & SERVICE_STATE_MASK;
+
+	switch (state) {
+	case SERVICE_STATE_ON:
+		/* Stay on if release leaves a client. */
+		if (srv->refs > 1U) {
+			notify = true;
+			rv = 1;
```

Comment:
Because I want to have `refs` updates done after all the logic is complete, rather than hidden inside a conditional in the switch statement.  This makes it parallel to the start logic.